Python/codecs.c

```diff
@@ -702,55 +702,62 @@ PyObject *PyCodec_IgnoreErrors(PyObject *exc)
 
 PyObject *PyCodec_ReplaceErrors(PyObject *exc)
 {
-    Py_ssize_t start, end, i, len;
+    Py_ssize_t start, end;
 
     if (PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeEncodeError)) {
-        PyObject *res;
-        Py_UCS1 *outp;
-        if (PyUnicodeEncodeError_GetStart(exc, &start))
+        if (_PyUnicodeError_GetParams(exc, NULL, NULL, &start, &end, false) < 0) {
             return NULL;
-        if (PyUnicodeEncodeError_GetEnd(exc, &end))
-            return NULL;
-        len = end - start;
-        res = PyUnicode_New(len, '?');
-        if (res == NULL)
+        }
+        if (end <= start) {
+            goto oob;
+        }
+        Py_ssize_t len = end - start;
```

Comment:
Would it make sense for `_PyUnicodeError_GetParams` to get `len` as well, clipping to `>=0`?
Then perhaps `len==0` can just be handled by the main code paths? The `oob` case with `Py_CONSTANT_EMPTY_STR` seems rare, so not too relevant as an *optimization*.

Reply:
> Then perhaps len==0 can just be handled by the main code paths

Unfortunately `PyUnicode_New(len, Py_UNICODE_REPLACEMENT_CHARACTER);` would not have a Unicode kind of `PyUnicode_2BYTE_KIND`. 0-strings are always 1 byte kind so the assertion after would fail. As for the XML charref replace handler, we will anyway need another check on `end - start`. So I think we'll be using the clipped len with half the handlers.

Reply:
I worry that this leaves in a similar gotcha as what this series of patches is fixing: a condition that the caller needs to remember to check for.
The assert could grow a  `|| len == 0`.

But, that's a nitpick. If you disagree, go ahead and merge!

Reply:
I think changing the `assert` would look better. I'll check that tomorrow offline and decide on Friday what to do.

Reply:
I have a local PR ready for adding the clipped length as a possible output. I'll open it on Sunday (I have some IRL stuff the two following days) but I think it'd be probably cleaner.